include/tree_walker.h

```diff
@@ -0,0 +1,134 @@
+#pragma once
+
+#include <utility>
+
+#include "exceptions.h"
+#include "smt.h"
+
+namespace smt {
+/* vector of pairs holding terms and ints that gets used within visit in the
+ * vector to_visit as a means of storing what terms we need to visit along with
+ * each term's child number relative to its parent node or a "-1 flag" that
+ * indicates all of a node's children have already been visited.
+ */
+using TermPairVec = std::vector<std::pair<Term, int>>;
+/* used to map from terms to their occurrences, represented as a pair of the
+ * formula it is found in and the path indicating the terms placement in the
+ * formula it is find in. A term's path is built up as list of indices where
+ * each index is a child number telling which path to follow from the topmost
+ * node to the term's place in the formula represented as a tree. For example,
+ * in the formula x+1=2, each term and its respective path is: x+1=2: []; x+1:
+ * [0]; x: [0,0]; 1: [0,1]; 2: [1]
+ */
+using UnorderedTermPairMap =
+    std::unordered_map<Term, std::pair<Term, std::vector<int>>>;
+
+/** \enum
+ * Walker_Continue : rebuild the current term and continue
+ * Walker_Skip     : skip this term and all subterms
+ * Walker_Abort    : abort visiting
+ */
+enum TreeWalkerStepResult
```

Comment:
This is not necessary, but I think it might be nice to use the same enums since they're identical. One way to do this is to create a new header file `walkers.h` or something like that and put the enum there and include that file in both `identity_walker.h` and `tree_walker.h`.

Reply:
Talked on slack with Yoni and decided this should be delayed to a separate PR, so no action necessary. Resolving this conversation.